pymysqlreplication/tests/binlogfilereader.py

```diff
@@ -0,0 +1,150 @@
+'''Read binlog files'''
+import struct
+
+from pymysql.util import byte2int
+from pymysqlreplication import constants
+from pymysqlreplication.event import FormatDescriptionEvent
+from pymysqlreplication.event import QueryEvent
+from pymysqlreplication.event import RotateEvent
+from pymysqlreplication.event import XidEvent
+from pymysqlreplication.row_event import TableMapEvent
+from pymysqlreplication.row_event import WriteRowsEvent
+
+class SimpleBinLogFileReader(object):
+    '''Read binlog files'''
+
+    _expected_magic = b'\xfebin'
+
+    def __init__(self, file_path, only_events=None):
+        self._current_event = None
+        self._file = None
+        self._file_path = file_path
+        self._only_events = only_events
+        self._pos = None
+
+    def fetchone(self):
+        '''Fetch one record from the binlog file'''
+        if self._pos is None or self._pos < 4:
+            self._read_magic()
+        while True:
+            event = self._read_event()
+            self._current_event = event
+            if event is None:
+                return None
+            if self._filter_events(event):
+                return event
+
+    def truncatebinlog(self):
+        '''Truncate the binlog file at the current event'''
+        if self._current_event is not None:
+            self._file.truncate(self._current_event.pos)
+
+    def _filter_events(self, event):
+        '''Return True if an event can be returned'''
+        # It would be good if we could reuse the __event_map in
+        # packet.BinLogPacketWrapper.
+        event_type = {
+            constants.QUERY_EVENT: QueryEvent,
+            constants.ROTATE_EVENT: RotateEvent,
+            constants.FORMAT_DESCRIPTION_EVENT: FormatDescriptionEvent,
+            constants.XID_EVENT: XidEvent,
+            constants.TABLE_MAP_EVENT: TableMapEvent,
+            constants.WRITE_ROWS_EVENT_V2: WriteRowsEvent,
+        }.get(event.event_type)
+        return event_type in self._only_events
+
+    def _open_file(self):
+        '''Open the file at ``self._file_path``'''
+        if self._file is None:
+            self._file = open(self._file_path, 'rb+')
```

Comment:
This line opens the MySQL binary log for writing.  If the user running the test does not have the necessary permissions, this will fail.